corehq/apps/locations/urls.py

```diff
@@ -31,6 +31,8 @@
     url(r'^$', default, name='default_locations_view'),
     url(r'^list/$', LocationsListView.as_view(), name=LocationsListView.urlname),
     url(r'^location_search/$', LocationsSearchView.as_view(), name='location_search'),
+    url(r'^location_search_has_users_only/$', LocationsSearchView.as_view(
+        include_locations_with_no_users_allowed=False), name='location_search_has_users_only'),
```

Comment:
What would you think of naming this after the place it's used rather than what distinguishes it from the other endpoint?  We'll also soon be excluding already-assigned locations from the results, so it may see further tweaking.

Reply:
I see from a later commit that the plan is to switch between endpoints depending on the desired functionality.  Would it be possible to make this enpdoint specific to the location assignment widget and then move the feature flag check there?  Otherwise further changes would need to account for both endpoints.

Reply:
Ah, I see that it's still the same view in both cases, but with a different initialization parameter here, I was thinking it was a subclass for some reason.  I still think there may be a place for making a `UserLocationWidgetView` and putting the logic there rather than passing it around, but that will probably make more sense with the next bit of deviation from the standard use case than it does now.

Reply:
I'm struggling a bit to articulate what I'm thinking, but let me try.  Essentially, there's a generic "locations select widget" that is used in several different places.  This is one customization specific to user location assignment, and we've got a couple more customizations likely coming.  The way it's done here is to set a flag on the widget, which directs to a different URL, which sets a flag on the view, which sets a flag on the controller, which forks the behavior in the controller's parent class.  That's a lot of steps, making it perhaps challenging to see when/how/why the flag is set to what it is.

The alternative I'm imagining would be to switch from the generic widget to one that's specific to user location assignment.  Then it's clear in what context the behavior differs.  Perhaps that'd still use the same system of passing flags around, only the flag would be called "is user location assignment" or some such.  Then when we fork behavior, we could do it all under the same flag or subclass, making the scope under which those changes apply clearer.

Again I don't think that's necessarily better, certainly not with just this logic change.  Mostly leaving my thoughts here now because I suspect it'll start to be more compelling as we further fork the logic (though it's always hard to tell for sure without trying it).

From the article Simon shared this morning:

> When designing abstractions, do not prematurely couple behaviours that may evolve separately in the longer term.

Here the widget for picking locations for a user is coupled to the generic widget for finding locations on a project, and I'm thinking that they are starting to evolve separately, so we may want to decouple them.